### ISSUE TITLE
Allow calling impure functions in clocked equations/statements

### ISSUE THREAD

**Issue by otronarp** [modelica/ModelicaSpecification#3614]
In _12.3 Pure Modelica Functions_ there is a list of situations where it's allowed to call an impure function, it doesn't mention clocked equations/statements. Shouldn't it be allowed there as well?

```
model M
  Clock c = Clock(1);
  Real x;
  Real y;
equation
  x = sample(time, c);
  y = impureFunction(x);
end M;
```

Feels a bit unnecessary to require the `y = ...` to be wrapped with `when Clock() then`.

**Comment by HansOlsson** [comment]
On one hand I see that point, but on the other hand I see a problem with usability and diagnostics.

Consider:
```
model M
  input Real x;
  output Real y;
equation
  y = impureFunction(x);
end M;

model A
  M m(x=sample(time, Clock(1));
end A;
```
Here the call is ok in `A` - but in `M` it would not be legal as it is not clocked. Assuming we want to allow the first model, the question is what to do with `M`:
- Say that it is illegal and have some weird rule that it must be clocked in the same model. (Not clear how.)
- Say that it is legal in `A` and give diagnostics for `M` ignoring that possibility. (Users of the feature will complain.)
- Say that it is legal in `A` and give diagnostics for `M` including that possibility, which will mean that every use of impure functions we have to include it.

Based on that I find the idea of explicitly making it clocked
```
model M
  input Real x;
  output Real y;
equation
  when Clock() then
    y = impureFunction(x);
  end when;
end M;
```
an acceptable way of writing to avoid all of that.

**Comment by HansOlsson** [comment]
As follow-up, having some other way of making models/blocks clocked is also something we could consider, but we would then need to differentiate between a model being clocked, vs. clocked with _one_ clock.

**Comment by otronarp** [comment]
I your example I would say `M` is illegal therefore `A` is also illegal.

**Comment by HansOlsson** [comment]
> I your example I would say `M` is illegal therefore `A` is also illegal.

But if they are illegal, I find it difficult to say that the equations are legal "just" because it is part of the same model (I realize that the discussion would benefit from having different names for the different models :-)

**Comment by HansOlsson** [comment]
@otronarp can we close this one without any action. I admit it could be clearer, but I don't see any good clean solution.

**Comment by otronarp** [comment]
>Based on that I find the idea of explicitly making it clocked
>...
>an acceptable way of writing to avoid all of that.

Regarding that, it works as a workaround for equations but not for algorithms since you cannot make them explicitly clocked, from _16.6 Clocked When-Clause_:

>The clocked when-clause cannot be nested and does not have any elsewhen part. It cannot be used inside an algorithm. General equations are allowed in a clocked when-clause.

So unless I'm missing something there is no way to call an `impure` function from a clocked algorithm, which to me feels like an unfortunate limitation.

**Comment by HansOlsson** [comment]
> > Based on that I find the idea of explicitly making it clocked
> > ...
> > an acceptable way of writing to avoid all of that.
> 
> Regarding that, it works as a workaround for equations but not for algorithms since you cannot make them explicitly clocked, from _16.6 Clocked When-Clause_:
> 
> > The clocked when-clause cannot be nested and does not have any elsewhen part. It cannot be used inside an algorithm. General equations are allowed in a clocked when-clause.
> 
> So unless I'm missing something there is no way to call an `impure` function from a clocked algorithm, which to me feels like an unfortunate limitation.

Agreed, I didn't consider algorithms earlier. 
This is especially problematic as it seems more natural to call `impure` functions from algorithms than from equations.

It still feels weird to require that impure function calls in algorithms can only be checked after base-clock partitioning.
A related weirdness is that one cannot say that an algorithm must be clocked without specifying the clock.

It may be that we should add some other way to make an algorithm/equation explicitly clocked.

**Comment by HansOlsson** [comment]
An example of a way to make an algorithm explicitly clocked would be to allow clocked when-clause in an algorithm iff it is the entire algorithm. Such as:
```Modelica
model MClockedAlg
  input Real x;
  output Real y;
algorithm
  when Clocked() then
    impureFunction1(x);
    y:=impureFunction2();
  end when;
end MClockedAlg;

model A
  MClockedAlg m(x=sample(time, Clock(1));
end A;
```

**Comment by otronarp** [comment]
Makes sense to me, but I lack insight into the details of the synchronous features to know if there are any potential pitfalls with that.
